for the search on reddit i started by a simple query on the r/programming subreddit

https://www.reddit.com/r/programming/search/?q=generative+ai+&type=link&cId=7fe9ba9a-76ca-41dc-9bac-2c3adbdcdb91&iId=41e374f5-d173-4905-910f-225ec7469651&sort=top

after that i asked chat gpt to create a script to retrieve the posts from the query

In [3]:

import praw

# Initialize Reddit API
reddit = praw.Reddit(
    client_id='vs9GwSkej8Y2SVB85OFIzw',
    client_secret='u3qYuCO1ULJEZj_HYw2DG7_yDi1z0A',
    user_agent='strategic project u/keepuru'
)

# Define the subreddit and search query
subreddit = reddit.subreddit('programming')
query = 'generative ai'

# Search for posts matching the query, sorted by score
search_results = subreddit.search(query, sort='top', limit=10000)  # Adjust limit as needed

# Dictionary to store results
results_dict = {}

# Store the titles, text of the post, and text of top comments in a dictionary
for submission in search_results:
    post_data = {
        "Title": submission.title,
        "Post Text": submission.selftext,  # Text of the post
        "Comments": []
    }
    submission.comments.replace_more(limit=None)  # Retrieve all comments, even nested ones
    top_comments = sorted(submission.comments, key=lambda x: x.score, reverse=True)[:1000]  # Sort comments by score and retrieve the top 1000
    for comment in top_comments:
        post_data["Comments"].append(comment.body)  # Text of the comment
    results_dict[submission.id] = post_data

the script above retrieves the top 10000 posts from r/programming when the generative ai query is given

the script also retrieves the text of the top 1000 comments from the post

In [5]:
len (results_dict)

208

unfortunately there are only 208 posts as showed in the cell above

In [4]:
import json
with open ("reddit posts generative ai.json", "w") as f:
    json.dump(results_dict, f, indent=4)

after the results get saved on a json file

after this i want to retrieve the same amount of information but for when the query is gen ai

In [6]:

import praw

# Initialize Reddit API
reddit = praw.Reddit(
    client_id='vs9GwSkej8Y2SVB85OFIzw',
    client_secret='u3qYuCO1ULJEZj_HYw2DG7_yDi1z0A',
    user_agent='strategic project u/keepuru'
)

# Define the subreddit and search query
subreddit = reddit.subreddit('programming')
query = 'gen ai'

# Search for posts matching the query, sorted by score
search_results = subreddit.search(query, sort='top', limit=10000)  # Adjust limit as needed

# Dictionary to store results
results_dict = {}

# Store the titles, text of the post, and text of top comments in a dictionary
for submission in search_results:
    post_data = {
        "Title": submission.title,
        "Post Text": submission.selftext,  # Text of the post
        "Comments": []
    }
    submission.comments.replace_more(limit=None)  # Retrieve all comments, even nested ones
    top_comments = sorted(submission.comments, key=lambda x: x.score, reverse=True)[:1000]  # Sort comments by score and retrieve the top 1000
    for comment in top_comments:
        post_data["Comments"].append(comment.body)  # Text of the comment
    results_dict[submission.id] = post_data

In [7]:
import json
with open ("reddit posts gen ai.json", "w") as f:
    json.dump(results_dict, f, indent=4)

not that many results from those

since the results i decided to query different subreddits and get the same results

here follows a generalization of the code above to swiftly query subreddits

In [15]:

import praw
from os import path
import os

# Initialize Reddit API
reddit = praw.Reddit(
    client_id='vs9GwSkej8Y2SVB85OFIzw',
    client_secret='u3qYuCO1ULJEZj_HYw2DG7_yDi1z0A',
    user_agent='strategic project u/keepuru'
)

def querySubreddit (subredditName:str, query:str):
    jsonFilePath = "reddit posts " + subredditName + " " + query + ".json"
    if not path.exists(jsonFilePath):
        print (jsonFilePath)
        # Define the subreddit and search query
        subreddit = reddit.subreddit(subredditName)
        query = query

        # Search for posts matching the query, sorted by score
        search_results = subreddit.search(query, sort='top', limit=10000)  # Adjust limit as needed

        # Dictionary to store results
        results_dict = {}

        # Store the titles, text of the post, and text of top comments in a dictionary
        for submission in search_results:
            post_data = {
                "Title": submission.title,
                "Post Text": submission.selftext,  # Text of the post
                "Comments": []
            }
            submission.comments.replace_more(limit=None)  # Retrieve all comments, even nested ones
            top_comments = sorted(submission.comments, key=lambda x: x.score, reverse=True)[:1000]  # Sort comments by score and retrieve the top 1000
            for comment in top_comments:
                post_data["Comments"].append(comment.body)  # Text of the comment
            results_dict[submission.id] = post_data

            
        with open (jsonFilePath, "w") as f:
            json.dump(results_dict, f, indent=4)
            
subreddits = ["programming", "AskProgramming", "learnprogramming", "ProgrammerHumor", "ProgrammingBuddies"]
queries = ["generative ai", "gen ai", "generative models", "machine-generated"]

for subreddit in subreddits:
    for query in queries:
        querySubreddit(subreddit, query)

reddit posts learnprogramming generative models.json


after this the next step is to merge all the text retrieved and apply the same logic we applied for the twitter dataset

In [ ]:
redditPosts = []
for subreddit in subreddits:
    for query in queries:
        with open ("reddit posts " + subreddit + " " + query + ".json") as f
            redditPosts.append(dict(json.load(f)))

allRedditPosts = {}
for redditPost in redditPosts:
    allRedditPosts.update(redditPost)
